### PR TITLE
tailcfg: add start of SSH policy to be sent from control plane to nodes

### PR DESCRIPTION
**Pull request by bradfitz**
Updates #3802

### CODE REVIEW

File: tailcfg/tailcfg.go [hunk 1]
```diff
@@ -1369,3 +1373,59 @@ type SetDNSRequest struct {
...
+	// match. If it does, the map entry's value is the actual
+	// user that's logged in.
+	SSHUser map[string]string `json:"sshUser"`
+
+	// Action is the outcome to task.
+	// A nil or invalid action means to deny.
+	Action *SSHAction `json:"action"`
+}
+
+// SSHPrincipal is either a particular node or a user on any node.
```
Comment (maisem) [inline tailcfg/tailcfg.go:1405]: What about tags? Do you intend on solving that by using `StableNodeID`?

File: tailcfg/tailcfg.go [hunk 2]
```diff
@@ -1369,3 +1373,59 @@ type SetDNSRequest struct {
...
+// An SSH rule is a match predicate and associated action for an incoming SSH connection.
+type SSHRule struct {
+	// Principal matches an incoming connection. If the connection
+	// matches anything in this list and also matches SSHUser,
+	// then Action is applied.
+	Principal []*SSHPrincipal `json:"principal"`
+
+	// SSHUser are the SSH users that this rule matches.
+	// It is a map from either ssh-user|"*" => local-user.
+	// The map must contain a key for either ssh-user or "*" to
```
Comment (maisem) [inline tailcfg/tailcfg.go:1395]: what if both match? I assume the `ssh-user` match takes precedence? Add comment?

File: tailcfg/tailcfg.go [hunk 3]
```diff
@@ -1369,3 +1373,59 @@ type SetDNSRequest struct {
...
+	User UserID       `json:"user,omitempty"`
+}
+
+// SSHAction is how to handle an incoming connection.
+// At most one field should be non-zero.
+type SSHAction struct {
+	// Message, if non-empty, is shown to the user before the
+	// action occurs.
+	Message string
+
```
Comment (maisem) [inline tailcfg/tailcfg.go:1418]: This deviates from the no deny policy in the ACLs, and as you mention elsewhere makes the ordering of rules matter. Can we simplify it and only have Accepts?

File: tailcfg/tailcfg.go [hunk 4]
```diff
@@ -1157,6 +1157,10 @@ type MapResponse struct {
 	// sees.
 	Health []string `json:",omitempty"`
 
+	// SSHPolicy, if non-nil, updates the SSH policy for how incoming
```
Comment (maisem) [inline tailcfg/tailcfg.go:1160]: update `CurrentMapRequestVersion`?

File: tailcfg/tailcfg.go [hunk 5]
```diff
@@ -1369,3 +1373,59 @@ type SetDNSRequest struct {
 	// Value is the value to add.
 	Value string
 }
+
+// SSHPolicy is the policy for how to handle incoming SSH connections
+// over Tailscale.
+type SSHPolicy struct {
+	// Rules are the rules to process for an incoming SSH
+	// connection. The first matching rule takes its action and
```
Comment (mayakacz) [inline tailcfg/tailcfg.go:1381]: Does this mean that if I have multiple rules, they're evaluated in order? I guess that makes sense. If we're allowing you to specify multiple SSH users though, and want to default to the non-root one, but you list them in a different order, does that mean we might use the root one instead?

File: tailcfg/tailcfg.go [hunk 6]
```diff
@@ -1369,3 +1373,59 @@ type SetDNSRequest struct {
...
+
+// An SSH rule is a match predicate and associated action for an incoming SSH connection.
+type SSHRule struct {
+	// Principal matches an incoming connection. If the connection
+	// matches anything in this list and also matches SSHUser,
+	// then Action is applied.
+	Principal []*SSHPrincipal `json:"principal"`
+
+	// SSHUser are the SSH users that this rule matches.
+	// It is a map from either ssh-user|"*" => local-user.
```
Comment (mayakacz) [inline tailcfg/tailcfg.go:1394]: I think this is saying we'll pass through the user that is authenticated in Tailscale? If so, yup.

File: tailcfg/tailcfg.go (hunk 5 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go:1381]: Yeah, but keep in mind that this SSHPolicy is not the user facing bits at all. This is the internal representation that tailscaled can handle, analogous to how ACL JSON compiles down to the internal PacketFilter/FilterRule:

https://pkg.go.dev/tailscale.com/tailcfg#MapResponse.PacketFilter
https://pkg.go.dev/tailscale.com/tailcfg#FilterRule

So this is just setting up tailscaled to be able to do configurable sets of things from control where we can implement higher level policies and UIs that map down to this.

File: tailcfg/tailcfg.go (hunk 3 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go:1434]: Added a comment about why it's needed.

File: tailcfg/tailcfg.go (hunk 2 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go]: done

File: tailcfg/tailcfg.go (hunk 1 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go:1417]: Yes, we can send the whole set down.

File: tailcfg/tailcfg.go (hunk 4 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go:1160]: Not yet. Once it's wired up to anything useful I can.

**Comment by bradfitz** [comment]
PTAL; ready to be submitted.

Still not wired up to anything, though. That'll be the next commit.

**Review by maisem — APPROVED** [review]

File: tailcfg/tailcfg.go [hunk 7]
```diff
@@ -1157,6 +1157,10 @@ type MapResponse struct {
 	// sees.
 	Health []string `json:",omitempty"`
 
+	// SSHPolicy, if non-nil, updates the SSH policy for how incoming
+	// SSH connections should be handled.
+	SSHPolicy *SSHPolicy
```
Comment (crawshaw) [inline tailcfg/tailcfg.go:1162]: omitempty?

File: tailcfg/tailcfg.go [hunk 8]
```diff
@@ -1369,3 +1373,82 @@ type SetDNSRequest struct {
...
 	Value string
 }
+
+// SSHPolicy is the policy for how to handle incoming SSH connections
+// over Tailscale.
+type SSHPolicy struct {
+	// Rules are the rules to process for an incoming SSH
+	// connection. The first matching rule takes its action and
+	// stops processing further rules.
+	Rules []*SSHRule
```
Comment (crawshaw) [inline tailcfg/tailcfg.go:1383]: given you're naming fields inside SSHRule, name this "rules"?

File: tailcfg/tailcfg.go [hunk 9]
```diff
@@ -1369,3 +1373,82 @@ type SetDNSRequest struct {
...
+	// Action is the outcome to task.
+	// A nil or invalid action means to deny.
+	Action *SSHAction `json:"action"`
+}
+
+// SSHPrincipal is either a particular node or a user on any node.
+// At most one field should be non-zero specified.
+type SSHPrincipal struct {
+	Node StableNodeID `json:"node,omitempty"`
+	User UserID       `json:"user,omitempty"`
```
Comment (crawshaw) [inline tailcfg/tailcfg.go:1421]: We have a StableUserID on the server, I think we should add it to tailcfg and use it here. I'd rather not introduce a use for the old UserID here.

File: tailcfg/tailcfg.go [hunk 10]
```diff
@@ -1369,3 +1373,82 @@ type SetDNSRequest struct {
...
+	// at least during beta testing period, when people can
+	// temporarily write their own SSHPolicy JSON file.
+}
+
+// SSHAction is how to handle an incoming connection.
+// At most one field should be non-zero.
+type SSHAction struct {
+	// Message, if non-empty, is shown to the user before the
+	// action occurs.
+	Message string
```
Comment (crawshaw) [inline tailcfg/tailcfg.go:1433]: Message and Reject don't have custom json names

**Review by crawshaw — APPROVED** [review]

File: tailcfg/tailcfg.go (hunk 8 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go]: done

File: tailcfg/tailcfg.go (hunk 10 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go]: done

File: tailcfg/tailcfg.go (hunk 9 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go]: adjusted

File: tailcfg/tailcfg.go (hunk 7 above)
Comment (bradfitz) [reply tailcfg/tailcfg.go]: done

**Pull request merged**